system/system-threading.md

```diff
@@ -0,0 +1,389 @@
+sends the appropriate protocol response to the function call.
+
+TODO: a diagram is worth 1000 words! ;-)
+
+As with the system thread marshaling, when no threading is enabled the same sequence
+of events happens, but without hopping between threads.
+
+### Non-marshaled functions
+
+Many system functions, such as `network_listening()` `spark_cloud_connect()` are
+implemented by setting or retrieving flags. These functions were already asynchronous
+by design before system threading, and so they are not marshaled to the system thread - setting or retrieving
+these flags on the application thread is sufficient and avoids the overhead of
+marshaling the request to another thread.
+
+
+
+## Future Improvements
+
+### Avoid Busy-Waiting
+
+Some coding styles have the application thread busy-waiting for events, e.g.
+
+```
+WiFi.connect();
+while (!WiFi.ready());
+```
+
+Busy waiting uses up processing cycles unnecessarily and could be achieved
+by blocking the thread until the condition is satisfied.
+
+We have added `waitUntil(WiFi.connected);` in anticipation of this, although
+presently the implementation still uses busy waiting.
+
+
+### Running Cloud Functions on a separate thread
+
+We might allow the user to register Cloud functions to `Particle.subscribe` and
+`Particle.function` so that they execute on separate thread.
+
+This would require:
+
+- a new syntax to declare these functions run on a "function thread".
+- a new function thread active object to execute application callback functions
+- documentation, education that these functions behave more like ISRs and so
+should expect to run concurrently with application code
+
+The ideal syntax is to declare the handler asynchronous:
+
+```
+Particle.subscribe("event", asynchronous(handler));
+```
+
+I don't know if this is feasible.
+
+This is not as nice, but definitely feasible:
+
+```
+Particle.subscribe(CALL_ASYNCHRONOUS, "event", handler);
+```
+
+
+### Provide "I don't care" call for functions returning results.
+
+`Particle.subscribe()` returns a true/false success result. Most code is written
+that doesn't check this result, so there's little point waiting for it, particularly
+as that waiting will block the application thread while the system connects to the network/cloud.
+
+A couple of solutions:
+
+1. add asynchronous versions for each function (not jazzed about this):
+
+```
+Particle.subscribeAsync("event", theHandler);
+```
+
+or
+
+```
+Particle.subscribe("event", theHandler, ASYNC);
+```
+
+(not to be confused with the use of CALL_ASYNCHRONOUS in the previous section - here the registration function
+is asynchronous, while the previous section discussed calling the registered function asynchronously.)
+
+
+2. Have a generic way to run code asynchronously on the system thread
+
+```
+ON_SYSTEM_THREAD(
+   Particle.subscribe("event", theHandler);
+   Particle.subscribe("metoo", anotherHandler);
+);
+```
+
```

Comment:
Ah, understood. Yes for this specific method that is a possibility, but it won't save much. The issue is that the system thread could be busy servicing the network, so any function calls that execute on the system thread will block until the system thread is free to service them. So even just adding the event to the local store will block. The alternative is to make the local event store thread-safe so that it can be updated from non-system threads, but that's something I'd rather not do since it introduces a lot more complexity.

In this case, I feel the caller can simply request the entire Particle.subscribe() to run on the system thread as I outlined int he last example.